System/Reader.h

```diff
@@ -1,502 +1,297 @@
+		/// <returns>Whether this Reader's stream is OK or not.</returns>
+		bool IsOK() { return m_pStream && m_pStream->good(); }
+
+		/// <summary>
+		/// Makes an error message box pop up for the user that tells them something went wrong with the reading, and where.
+		/// </summary>
+		/// <param name="errorDesc">The message describing what's wrong.</param>
+		void ReportError(std::string errorDesc);
+#pragma endregion
+
+	protected:
+
+		/// <summary>
+		/// A struct containing information from the currently used stream.
+		/// </summary>
+		struct StreamInfo {
+			StreamInfo(std::ifstream *pStream, std::string filePath, int currentLine, int prevIndent) :
+				m_pStream(pStream), m_FilePath(filePath), m_CurrentLine(currentLine), m_PreviousIndent(prevIndent) { ; }
+
+			// Owned by the reader, so not deleted by this
+			std::ifstream *m_pStream; //! Currently used stream, is not on the StreamStack until a new stream is opened.
+			std::string m_FilePath; //! Currently used stream's filepath.
+			int m_CurrentLine; //! The line number the stream is on.
+			int m_PreviousIndent; //! Count of tabs encountered on the last line Eat() ate.
+		};
+
+		static const std::string ClassName; //! A string with the friendly-formatted type name of this.
+
+		std::ifstream *m_pStream; //! Currently used stream, is not on the StreamStack until a new stream is opened.
+		std::list<StreamInfo> m_StreamStack; //! Stack of stream and filepath pairs, each one representing a file opened to read from within another.
+		bool m_EndOfStreams; //! All streams have been depleted.
+
+		void(*m_fpReportProgress)(std::string, bool); //! Function pointer to report our reading progress to, by calling it and passing a descriptive string to it.
+
+		std::string m_FilePath; //! Currently used stream's filepath.
+		std::string m_FileName; //! Only the name of the currently read file, excluding the path.
+
+		std::string m_DataModuleName; //! The current name of the data module being read from, including the .rte extension.
+		int m_DataModuleID; //! The current ID of the data module being read from.
+
+		int m_PreviousIndent; //! Count of tabs encountered on the last line Eat() ate.
+		int m_IndentDifference; //! Difference in indentation from the last line to the current line.
+		std::string m_ReportTabs; //! String containing the proper amount of tabs for the report.
+
+		int m_CurrentLine; //! The line number the stream is on.
+		bool m_OverwriteExisting; //! Whether object instances read from this should overwrite any already existing ones with the same names.
+		bool m_SkipIncludes; //! Indicates whether reader should skip included files
+
+		/// <summary>
+		/// When NextProperty() has returned false, indicating that there were no more properties to read on that object,
+		/// this is incremented until it matches -m_IndentDifference, and then NextProperty will start returning true again.
+		/// </summary>
+		int m_ObjectEndings;
+
+#pragma region Reading Operations
```

Comment:
Because we agreed members come before methods and it applies everywhere.

> I'm actually totally in favour of having members be above methods cause that's how most languages do it, but I don't mind having them below.

Reply:
Ooops. Sorry, my brain was melting at that point. I caught myself later somewhere when I realized we'd planned to do it this way.